### PR TITLE
Fix the bug in THCTensor_(baddbmm) and ATen's addmm_cuda for strided views input

### LINKED ISSUE #42418: einsum fails in THCudaBlas_DgemmStridedBatched only when running on GPU
## 🐛 Bug

I called `torch.einsum` with manipulated strides. And got the error (Two errors, because I tested in different environments):
```
RuntimeError: cublas runtime error : an invalid numeric value was used as an argument at /pytorch/aten/src/THC/THCBlas.cu:281  # torch.__version__ 1.7.0.dev20200705+cu101
RuntimeError: cublas runtime error : an invalid numeric value was used as an argument at /pytorch/aten/src/THC/THCBlas.cu:359  # torch.__version__ 1.5.1
```

## To Reproduce
I did some memory optimizations for https://github.com/fgnt/nara_wpe/pull/47/files .
To keep it simple in this issue, I produced a toy example that also fails:

```python
import torch
print('torch.__version__', torch.__version__)

x = torch.tensor([[1., 2, 3], [4., 5, 6]]).to(0)

# Do segment_axis
# https://github.com/fgnt/nara_wpe/blob/ae0fd6444a6bb03aa6e29ea2b698e6b73d596253/nara_wpe/wpe.py#L14
# This is a view on the input data, but with a segmentation that is used for example in the stft.
# This view can save large memory.
stride = list(x.stride())
stride.insert(1, 1)
shape = list(x.size())
shape.insert(1, 2)
shape[-1] -= 1
c = torch.as_strided(x, size=shape, stride=stride)

print(c.stride(), c.shape)

torch.einsum('...ab,...bc->...ac', c, c)
```

The stdout on one computer is:
```
torch.__version__ 1.7.0.dev20200705+cu101
(3, 1, 1) torch.Size([2, 2, 2])
 ** On entry to SGEMM  parameter number 8 had an illegal value
...
RuntimeError: cublas runtime error : an invalid numeric value was used as an argument at /pytorch/aten/src/THC/THCBlas.cu:281
```

and on another:

```
torch.__version__ 1.5.1
(3, 1, 1) torch.Size([2, 2, 2])
...
RuntimeError: cublas runtime error : an invalid numeric value was used as an argument at /pytorch/aten/src/THC/THCBlas.cu:359
```

## Expected behavior

That the toy example works.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.5.1 and 1.7.0.dev20200705+cu101
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: 11.0 and 10.X (I do not know the exact cuda value, probably 10.1)
 - GPU models and configuration: GTX 980 and GTX1080
 - Any other relevant information:

## Additional context

cc @ngimel @mruberry @rgommers @vincentqb @vishwakftw @SsnL @jianyuh

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Fixes #42418. 

The problem was that the non-contiguous batched matrices were passed to `gemmStridedBatched`.

The following code fails on master and works with the proposed patch:
```python
import torch
x = torch.tensor([[1., 2, 3], [4., 5, 6]], device='cuda:0')
c = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
torch.einsum('...ab,...bc->...ac', c, c)
```

### CODE REVIEW

**Comment by ngimel** [comment]
Thank you, the fix looks good! Please add the test to the test suite. There's test_baddbmm in test_torch.py, but for some reason it's enabled only on the CPU. Can you try enabling it on cuda and adding a testcase for the behavior your enabled there?

**Comment by IvanYashchuk** [comment]
Sure, I will do that.

**Comment by IvanYashchuk** [comment]
While writing tests I found that `torch.mm` also doesn't work with this kind of strided input
```python
import torch
x = torch.tensor([[1., 2, 3], [4., 5, 6]], device='cuda:0')
c = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
torch.mm(c[0], c[0]) # Fails
```
The same problem was in `ATen/native/cuda/LinearAlgebra.cu:prepare_matrix_for_cublas` as in `THCTensor_(baddbmm)` that the input was not transformed into a contiguous array. I have fixed that.
I've added the tests both for `pytorch.mm` and `pytorch.bmm`. Correctness is checked via comparing to the NumPy result.

File: test/test_torch.py [hunk 1]
```diff
@@ -17035,6 +17035,32 @@ def genf_float(x, y):
...
 
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float32, torch.float64)
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_strided_mm_bmm(self, device, dtype):
+        # Tests strided view case with stride smaller than corresponding dimension size
+        x = torch.tensor([[1., 2., 3.], [4., 5., 6.]], dtype=dtype, device=device)
+        new_shape = [2, 2, 2]
+        new_stride = [3, 1, 1]
+        sx = torch.as_strided(x, size=new_shape, stride=new_stride)
```
Comment (ngimel) [inline test/test_torch.py:17046]: can you please use `compare_with_numpy` helper here to compare results? You don't need numpy arrays to have the same striding, so `.detach().cpu().numpy()` that `compare_with_numpy` does should be fine. To deal with multiple arguments, you can do (for torch_fn, and similarly for numpy_fn) `lambda x: torch.bmm(x,x)`. 
You don't need to compare results of torch.bmm with torch.mm, if you are comparing with numpy anyway

**Review by ngimel — COMMENTED** [review]
Thank you, this looks great! I have a small suggestion about the test.

File: test/test_torch.py (hunk 1 above)
Comment (IvanYashchuk) [reply test/test_torch.py:17046]: I've modified the test to use `compare_with_numpy` now.

**Review by ngimel — APPROVED** [review]
Awesome, thanks!

**Comment by ngimel** [comment]
flake8 error is real

**Comment by IvanYashchuk** [comment]
> flake8 error is real

Is it okay to ignore E731 because of assigned lambdas instead of def in the test?

**Comment by ngimel** [comment]
Yeah, that's fine, you could also use lambdas directly as arguments, but it does not matter. Sorry to ask, but can you please rebase? We had an issue with docker images today, so CI on this PR is failing because it is against a bad base, and can't find docker images.

**Comment by IvanYashchuk** [comment]
Alright, I did something wrong here. I've rebased onto master and pushed to branch and PR was automatically closed.

**Comment by ngimel** [comment]
It somehow became 0-commit 0-line pull request, that's probably why it was closed. Something wrong with the rebase?

**Comment by IvanYashchuk** [comment]
I've recovered the branch. Can this PR be reopened? I obviously don't have rights for this. I am sorry for the inconvenience.

**Comment by ngimel** [comment]
Thank you!

**Pull request closed without merging**